File: backend/app/api/v1/groups.py
"""人员组 (people groups) — a company-wide shared variable pool.

Moved out of the admin backend: any employee holding the ``feature:group``
permission may view and edit groups. Group members are consumed
by dispatch / member pickers via ``/users/groups`` (open to all active users).
"""

from fastapi import APIRouter, Depends, Response, status
from sqlalchemy.orm import Session

from app.api.v1.deps import csrf_protect, require_permission
from app.core.permissions import FEATURE_GROUP
from app.db.session import get_db
from app.models.user import User
from app.schemas.admin import (
    GroupCreate,
    GroupFromDepartmentIn,
    GroupImportSourceOut,
    GroupMembersIn,
    GroupOut,
    GroupUpdate,
)
from app.services.admin_service import AdminService

router = APIRouter(prefix="/groups", tags=["groups"])


@router.get("", response_model=list[GroupOut])
def list_groups(
    db: Session = Depends(get_db),
    user: User = Depends(require_permission(FEATURE_GROUP)),
) -> list:
    return AdminService(db, user).list_groups()


@router.get("/import-sources", response_model=list[GroupImportSourceOut])
def list_import_sources(
    db: Session = Depends(get_db),
    user: User = Depends(require_permission(FEATURE_GROUP)),
) -> list:
    # Departments as import candidates — id+name only, no admin:department needed.
    return [
        GroupImportSourceOut(id=d.id, name=d.name)
        for d in AdminService(db, user).list_departments()
    ]


@router.post("", response_model=GroupOut, dependencies=[Depends(csrf_protect)])
def create_group(
    payload: GroupCreate,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission(FEATURE_GROUP)),
):
    return AdminService(db, user).create_group(payload)


@router.post(
    "/from-department",
    response_model=GroupOut,
    dependencies=[Depends(csrf_protect)],
)
def create_group_from_department(
    payload: GroupFromDepartmentIn,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission(FEATURE_GROUP)),
):
    return AdminService(db, user).create_group_from_department(payload.department_id, payload.name)


@router.patch("/{group_id}", response_model=GroupOut, dependencies=[Depends(csrf_protect)])
def update_group(
    group_id: int,
    payload: GroupUpdate,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission(FEATURE_GROUP)),
):
    return AdminService(db, user).update_group(group_id, payload)


@router.post(
    "/{group_id}/members",
    response_model=GroupOut,
    dependencies=[Depends(csrf_protect)],
)
def set_group_members(
    group_id: int,
    payload: GroupMembersIn,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission(FEATURE_GROUP)),
):
    return AdminService(db, user).set_group_members(group_id, payload.member_ids)


@router.delete(
    "/{group_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    dependencies=[Depends(csrf_protect)],
)
def delete_group(
    group_id: int,
    db: Session = Depends(get_db),
    user: User = Depends(require_permission(FEATURE_GROUP)),
) -> Response:
    AdminService(db, user).delete_group(group_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/groups.ts
import { api } from './client';
import type { GroupOut } from './admin';

// 人员组共享变量池 —— 全公司共享，凭 feature:group 访问（owner 也按权限行判断）。
// 端点独立于后台管理（/groups），成员消费仍走 /users/groups。

export interface GroupInput {
  name: string;
  description?: string | null;
  sort_order?: number;
  member_ids?: number[];
}

export interface GroupImportSource {
  id: number;
  name: string;
}

export function listGroups(): Promise<GroupOut[]> {
  return api<GroupOut[]>('/groups');
}

export function listGroupImportSources(): Promise<GroupImportSource[]> {
  return api<GroupImportSource[]>('/groups/import-sources');
}

export function createGroup(input: GroupInput): Promise<GroupOut> {
  return api<GroupOut>('/groups', { method: 'POST', body: input });
}

export function updateGroup(
  id: number,
  input: Partial<Omit<GroupInput, 'member_ids'>>,
): Promise<GroupOut> {
  return api<GroupOut>(`/groups/${id}`, { method: 'PATCH', body: input });
}

export function setGroupMembers(id: number, memberIds: number[]): Promise<GroupOut> {
  return api<GroupOut>(`/groups/${id}/members`, {
    method: 'POST',
    body: { member_ids: memberIds },
  });
}

export function createGroupFromDepartment(
  departmentId: number,
  name?: string | null,
): Promise<GroupOut> {
  return api<GroupOut>('/groups/from-department', {
    method: 'POST',
    body: { department_id: departmentId, name: name ?? null },
  });
}

export function deleteGroup(id: number): Promise<null> {
  return api<null>(`/groups/${id}`, { method: 'DELETE' });
}
